tools/depcheck/pkg/graph/graph.go

```diff
@@ -55,7 +60,140 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
 	return nil
 }
 
+// RemoveNode deletes the given node from the graph,
+// removing all of its outbound and inbound edges as well.
+func (g *MutableDirectedGraph) RemoveNode(n *Node) {
+	delete(g.nodesByName, n.UniqueName)
+	g.DirectedGraph.RemoveNode(n)
+}
+
 func (g *MutableDirectedGraph) NodeByName(name string) (graph.Node, bool) {
 	n, exists := g.nodesByName[name]
 	return n, exists && g.DirectedGraph.Has(n)
 }
+
+// PruneOrphans recursively removes nodes with no inbound edges.
+// Receives a list of UniqueName prefixes of nodes to ignore.
+// Returns a list of recursively pruned nodes.
+func (g *MutableDirectedGraph) PruneOrphans(ignore []string) []*Node {
+	removed := []*Node{}
+
+	for _, n := range g.Nodes() {
+		node, ok := n.(*Node)
+		if !ok {
+			continue
+		}
+		if len(g.To(n)) > 0 {
+			continue
+		}
+
+		skip := false
+		for _, toIgnore := range ignore {
+			if strings.HasPrefix(node.UniqueName, toIgnore) {
+				skip = true
+				break
+			}
+		}
+		if skip {
+			continue
+		}
+
+		g.RemoveNode(node)
+		removed = append(removed, node)
+	}
+
+	if len(removed) == 0 {
+		return []*Node{}
+	}
+
+	return append(removed, g.PruneOrphans(ignore)...)
+}
+
+// FindExclusiveDependencies receives a set of targetNodes.
+// The graph is cloned, and all targetNodes are removed from it.
+// Returns the resulting set of orphaned nodes recursively pruned from the cloned graph.
+func (g *MutableDirectedGraph) FindExclusiveDependencies(targetNodes []*Node, ignore []string) []*Node {
```

Comment:
I don't think this should be base functionality of your graph.  Move this to an analysis package.